src/utils/helpers/DetailsDrawHelpers.ts

```diff
@@ -0,0 +1,223 @@
+        .attr('cursor', 'pointer')
+        .append('rect')
+        .attr('transform', 'translate(5,5)')
+        .attr('width', 10)
+        .attr('height', 10)
+        .attr('fill', ElementColors.DETAILS_BACKGROUND);
+    selectDetailsExitButtonWrapper()
+        .append('path')
+        .attr('transform', 'translate(10,10) rotate(45)')
+        .attr(
+            'd',
+            symbol()
+                .type(symbolCross)
+                .size(20)
+        );
+    createZoom(container, Selectors.ZOOM_DETAILS);
+    container.call(
+        zoom<any, any>().on('zoom', changeZoom(Selectors.ZOOM_DETAILS)).transform,
+        // rough center screen on diagram's root node
+        zoomIdentity.translate(-width / 5.3, -height / 2.65)
+    );
+}
+
+interface TreeStructure {
+    name: string;
+    children?: TreeStructure[];
+}
+
+type TreeNodeWithVisitedFlag = TreeNode & { isVisited?: boolean };
+
+function mapNodeToTreeStructure(node: TreeNodeWithVisitedFlag, links: (keyof TreeNode & 'consumers') | 'providers'): TreeStructure {
+    node.isVisited = true;
+    const UnvisitedLinks = node[links].filter((node: TreeNodeWithVisitedFlag) => !node.isVisited && node[links]);
+    const children = UnvisitedLinks.length > 0 ? UnvisitedLinks.map(nestedNode => mapNodeToTreeStructure(nestedNode, links)) : [];
+    node.isVisited = undefined;
+    return { name: node.name, children };
+}
+
+const VERTICAL_DISTANCE_BETWEEN_NODES = 40;
+const HORIZONTAL_DISTANCE_BETWEEN_NODES = 300;
+
+const NODE_NEIGHBOURS_SEPARATION_MULTIPLIER = 1;
+const NODE_NORMAL_SEPARATION_MULTIPLIER = 4;
+
+function createTree(data: TreeStructure) {
+    const node = hierarchy(data);
+    return tree<TreeStructure>()
+        .nodeSize([VERTICAL_DISTANCE_BETWEEN_NODES, HORIZONTAL_DISTANCE_BETWEEN_NODES])
+        .separation((node1, node2) =>
+            node1.parent === node2.parent ? NODE_NEIGHBOURS_SEPARATION_MULTIPLIER : NODE_NORMAL_SEPARATION_MULTIPLIER
+        )(node);
+}
+
+function getRootYPosition(data: HierarchyPointNode<TreeStructure>) {
+    let x0 = Infinity;
+    let x1 = -x0;
+    data.each(node => {
+        if (node.x > x1) x1 = node.x;
+        if (node.x < x0) x0 = node.x;
+    });
+    return VERTICAL_DISTANCE_BETWEEN_NODES - x0;
+}
+
+function createDiagram(
+    tree: HierarchyPointNode<TreeStructure>,
+    containerWidth: number,
+    containerHeight: number,
+    rootNodeXOffset: number,
+    drawToLeft?: boolean
+) {
+    if (!tree.children || !tree.children.length) {
+        return null;
+    }
+
+    const diagramWidth = containerWidth / 2;
+    const diagramXOffset = -diagramWidth / 8;
+    const diagramYOffset = -containerHeight / 2;
+
+    const svg = create('svg').attr('viewBox', `${diagramXOffset} ${diagramYOffset + rootNodeXOffset} ${diagramWidth} ${containerHeight}`);
+
+    const g = svg
+        .append('g')
+        .attr('font-size', 15)
+        .attr('transform', `translate(0,${rootNodeXOffset}) ${drawToLeft ? 'rotate(180)' : ''}`);
```

Comment:
could have function for returning that string

Reply:
Extracted transforming of diagram elements to a function